# SageMath: Uvod

## Kaj je SageMath (SageMath)?

- SageMath (na kratko Sage) je brezplačen odprtokodni programski sistem za matematiko.
- Lahko si ga predstavljate kot Phyton na matematičnih steroidih.
- Poslanstvo sistema SageMath je ustvariti učinkovito brezplačno odprtokodno alternativo sistemom Magma, Maple, Mathematica in Matlab.

**Glej tudi**
SageMath se pogosto uporablja in na spletu je na voljo veliko gradiva. Navajam nekaj najuporabnejših virov.

- SageMath Documentation: [https://doc.sagemath.org/](https://doc.sagemath.org/)
- SageMath Reference Manual: [https://doc.sagemath.org/html/en/reference/index.html](https://doc.sagemath.org/html/en/reference/index.html)
- A tour of SageMath: [https://doc.sagemath.org/html/en/a_tour_of_sage/](https://doc.sagemath.org/html/en/a_tour_of_sage/)
- Introductory SageMath tutorial: [https://doc.sagemath.org/html/en/prep/Intro-Tutorial.html](https://doc.sagemath.org/html/en/prep/Intro-Tutorial.html)
- SageMath tutorial: [https://doc.sagemath.org/html/en/tutorial/index.html](https://doc.sagemath.org/html/en/tutorial/index.html)
- SageMath thematic tutorial: [https://doc.sagemath.org/html/en/thematic_tutorials/index.html](https://doc.sagemath.org/html/en/thematic_tutorials/index.html)
- SageMath repository of interacts (interaktivni programčki): [https://wiki.sagemath.org/interact](https://wiki.sagemath.org/interact)
- SageMath quickref cards (reference za pomoč): [https://wiki.sagemath.org/quickref](https://wiki.sagemath.org/quickref)


## Kako uporabljamo SageMath?

SageMath lahko uporabljate na več načinov, vsak od njih ponuja različno raven nadzora in različne funkcije. Glavni načini uporabe so:

- **SageCell** [https://sagecell.sagemath.org/](https://sagecell.sagemath.org/) Zelo uporabno za enostavne in hitre naloge.

- **Lokalna namestitev**. Ni priporočljivo za začetnike. Če nimate izkušenj z nameščanjem programske opreme, je to lahko nekoliko zapleteno. Poleg tega skoraj nikoli ni potrebna. Edini dober razlog za lokalno namestitev programa SageMath je, da nimate dostopa do interneta.

- **Interaktivni programčki**. Podobno kot apleti GeoGebre. Za uporabo interaktivnih programčkov ni treba imeti predznanja. Za njihovo ustvarjanje pa potrebujete napredno poznavanje programa SageMath, ki presega vsebine tega predmeta.

- **CoCalc.com** [https://cocalc.com/](https://cocalc.com/). S preprostimi besedami, CoCal je za SageMath, kar je Overleaf za LaTeX.

## SageMath na CoCalcu

Obstajajo različne načine za uporabo SageMath, najpogosteje so:

- Na ukazni vrstici (terminalu). Ta način je primeren za napredne uporabnike in ni priporočljiv za začetnike. V tem primeru, SageMath deluje podobno kot običajen programski jezik, ukaze napišemo v terminal in dobimo rezultate. Ukazna vrstica običajno izgleda tako:



<img alt="terminal" src="./sagemath/terminal.png" width="500px" style="margin:auto;">

- Scripting: Kodo napišemo v besedilno datoteko in jo nato zaženete v ukazni vrstici. Ta način je prav tako primeren za napredne uporabnike in ni priporočljiv za začetnike, saj zahteva poznavanje osnov programiranja.



- **Interaktivno v zvezku (Jupyter Notebook)**. Ta način je najbolj primeren za začetnike in je tisti, ki ga bomo uporabljali v tem predmetu. SageMath lahko uporabljamo v zvezkih Jupyter, ki omogočajo kombiniranje besedila, matematičnih izrazov in kode v enem dokumentu. Zvezki so zelo uporabni za učenje, raziskovanje in dokumentiranje matematičnih izračunov. V tem predmetu bomo uporabljali CoCalc, spletno platformo, ki omogoča uporabo SageMath v zvezkih Jupyter brez potrebe po lokalni namestitvi.


**Vaja 1**
Pojdi na [https://cocalc.com/](CoCal.com) and ustvarite račun.


**Vaja 2**
Ustvarite nov Jupyter zvezek na CoCal-u z jedrom SageMath






::::{solution} ex-ustvariDatoteko
:class: tip dropdown

1. Na CoCal-u ustvarite nov datoteko.
:::{image} ./sagemath/jupyter1.png
:width: 400px
:align: center
:::

2. Poimenujete ga in izberite "Jupyter Notebook" (pazite, da izberite "Jupyter Notebook" in ne "SageMath Notebook").
:::{image} ./sagemath/jupyter2.png
:width: 400px
:align: center
:::


3. Izberite kernel SageMath 10.4 (ali najnovejšo različico).
:::{image} ./sagemath/jupyter3.png
:width: 400px
:align: center
::::



V beležnici Jupyter imamo dve vrsti celic: *Code* (koda) in *Text* (besedilo) 


<img alt="celice" src="./sagemath/jupyter4.png" width="500px" style="margin:auto;">


V celici z besedilo lahko uporabite _Markdown_ ali standardne besedilo. Priporočamo, da uporabite _Markdown_, saj omogoča boljšo oblikovanje besedila in vključevanje matematičnih izrazov.



Markdown je preprost jezik za označevanje besedila, ki omogoča enostavno oblikovanje besedila. Pristop je podoben LaTeX-u, vendar je veliko enostavnejši za uporabo, na žalost, pa tudi manj zmogljiv.

Osnovne značilnosti Markdowna vključujejo:

- Naslovi: Uporabite `# Naslov` za naslove. Več `#` pomeni nižjo raven naslova.
- Krepko in ležeče besedilo: Uporabite `**krepko**` za **krepko** besedilo in `_ležeče_` za _ležeče besedilo_.
- Seznami: Uporabite `-` ali `*` za neurejene sezname in številke za urejene sezname.
- Povezave: Uporabite `[besedilo](URL)` za ustvarjanje povezav.
- V Jupyter zvezkih lahko vključite tudi matematični izraze z uporabo LaTeX sintakse, na primer `$x^2 + y^2 = z^2$` za inline izraze in `$$E = mc^2$$` za prikazne izraze.


Za več informacij o Markdownu si oglejte naslednje vire: [Markdown osnovna sintaksa](https://www.markdownguide.org/basic-syntax/)

Ta učbenik je napisan v Myst Markdown, ki je razširitev Markdowna.

--- 

V celici z kodo vnesemo ukaze SageMath. Za izvedbo kode pritisnite {kbd}`Shift` + {kbd}`Enter` ali kliknite na gumb "Run" v orodni vrstici. Rezultati se bodo prikazali neposredno pod celico z kodo.

## SageMath kot kalkulator

SageMath lahko uporabljamo kot kalkulator za osnovne aritmetične operacije, kot so seštevanje, odštevanje, množenje in deljenje. Poleg tega lahko uporabljamo tudi bolj zapletene matematične funkcije in operacije.

Z nekaj manjšimi izjemami Sage uporablja programski jezik Python, zato so osnovne aritmetične operacije in druge osnove programiranja v SageMath zelo podobne tistim v Pythonu.


Poglejte naslednje primere. Še bolj pa jih preizkusite sami v svojem CoCalc zvezku!

In [60]:
1+1

2

In [61]:
( 1 + 2 * (3 + 5) ) * 2

34

Simbol * pomeni množenje, ki se ne sme izpustiti, tudi v izrazih, kot je $2x$: `2*x`. Za združevanje simbolov uporabljamo oklepaj `(` `)` nikoli `[` `]` ali `{` `}`.

In [62]:
2**3 # ** pomeni potenciranje, kot v Pythonu

8

In [63]:
2^3 # ampak ^ pomeni tudi potenciranje, za razliko od Pythona

8

In [64]:
-3^2 # pazite! 

-9

Vrstni red izvajanja operacij je določen v [Arithmetical binary operator precedence](https://doc.sagemath.org/html/en/tutorial/appendix.html#section-precedence).

In [65]:
20/6 # Deljenje se označi z /

10/3

Pri vzeto SageMath poskuša privzeto vrniti izraz, ki je čim bolj natančen. To je `10/3` vrne racionalno število $10/3$ (in ne decimalno približek, kot je $3,33333$). Če želimo številčni približek, lahko enemu od števil preprosto dodamo decimalno piko.

In [66]:
20.0/6

3.33333333333333

Kvadratne korene lahko izračunamo z ukazom `sqrt()`. Upoštevajte, da če je argument celo število, SageMath interpretira rezultat kot simbolno število.


In [67]:
sqrt(2)

sqrt(2)

In [68]:
sqrt(8)

2*sqrt(2)

Seveda, lahko dobimo številčni približek z ukazom `sqrt(8.0)`. Na splošno lahko dobimo numerične približke izrazov z (enakovrednimi) ukazi `n()`, `N()`, `numerical_approx()`. Vsi zgornji ukazi imajo neobvezen argument `digits` za določitev natančnosti približke.


In [69]:
n(sqrt(8), digits=50)

2.8284271247461900976033774484193961571393437507539

V [Tabeli %s](#tab-operacije) so povzete osnovne aritmetične operacije v SageMath-u.

:::{list-table} Osnovne aritmetične operacije v SageMath-u
:header-rows: 0
:label: tab-operacije

* - Seštevanje ($a+b$)
  - `a+b` 
* - Odštevanje ($a-b$)
  - `a-b`
* - Množenje ($ab$)
  - `a*b` 
* - Deljenje ($\frac{a}{b}$)
  - `a/b`
* - Potenciranje ($a^b$)
  - `a**b` ali `a^=b` 
* - Kvadratni koren ($\sqrt{a}$)
  - `sqrt(a)`
* - Splošni koren ($\sqrt[n]{a}$)
  - `a^(1/n)`
:::

:::{exercise}
:label: ex-racunSageMath
Z uporabo LaTeX-a napišite rešitve vsake vaje v ločeno celico besedila.

1. Deli $28$ z $2$, povišanim na $5$. potenco, kot racionalno število, nato pa dobi njegov desetiški približek točke, kjer so vse cifre pomembne.
1. Izračunajte decimalni približek števila $\sqrt{2}$.
1. S programom SageMath izračunajte $(-9)^{1/2}$. Opišite rezultat.
:::



::::{solution} ex-racunSageMath
:class: tip dropdown



In [70]:
28/(2**5)

7/8

n(28/(2**5), digits=3)

Ker je $$\frac{28}{32} = 0.875,$$ je njegov decimalni približek z vsemi pomembnimi ciframi $0.875$.



In [71]:
sqrt(2).n()

1.41421356237310

Decimalni približek števila $\sqrt{2}$ je približno $1.41421356237310$.

In [72]:
(-9)^(1/2)

3*I

Rezultat je $3i$, kjer je $i$ imaginarna enota, saj je koren iz negativnega števila v kompleksnih številih definiran kot produkt korena iz pozitivnega števila in imaginarne enote $i$. V SageMath-u je imaginarna enota predstavljena z `I`.


::::

## Operacije s celimi števili


S simboli `//` in `%` dobimo (celoštevilski) kvocient oziroma ostanek dveh celih števil


$$ 20 = 6 \cdot 3 + 2$$


In [73]:
20 // 6

3

In [74]:
20 % 6

2

In [75]:
divmod(20,6) # vrne kvocient in ostanek hkrati kot tuple

(3, 2)

Izračunamo lahko tudi fakultete in binomske koeficiente

$$3!=6$$


In [76]:
factorial(3)

6

$$ \binom{4}{2} = 6 $$


In [77]:
binomial(4,2)

6

Število lahko faktoriziramo kot produkt praštevil


In [78]:
factor(factorial(7))

2^4 * 3^2 * 5 * 7

Lahko uporabimo naslednji način tudi


In [79]:
(factorial(7)).factor()

2^4 * 3^2 * 5 * 7

Če želimo vedeti, ali je število praštevilo, lahko uporabimo ukaz `is_prime()`:

In [80]:
is_prime(29)

True

In [81]:
is_prime(65537)

True

Delitelje števila lahko dobimo na naslednji način.

In [82]:
12.divisors()

[1, 2, 3, 4, 6, 12]

Če samo želimo delitelje, ki so praštevila:


In [83]:
120.prime_divisors()

[2, 3, 5]

In [84]:
120.prime_factors()

[2, 3, 5]

Izračunamo lahko _največji skupni delitelj_ in _najmanjši skupni večkratnik_ dveh števil.


$$D(120,48) = $$


In [85]:
gcd(120,48)

24

$$v(9,15) = $$


In [86]:
lcm(9,15)

45

:::{exercise}
:label: ex-celaStevilaSageMath
Z uporabo SageMath-a rešite naslednje naloge, rešitve napišite v ločeno celico besedila z uporabo LaTeXa.

1. Poiščite ročno količnik in ostanek pri deljenju 956 z 98.
1. S programom SageMath preverite, ali sta zgoraj izračunana količnik in ostanek pravilna.
1. S programom SageMath ugotovite, ali $3$ deli $234878$.
1. Izračunajte seznam deliteljev za vsako od celih števil $134$, $491$, $422$ in $1002$.
1. Katera od zgornjih celih števil so praštevilska?
1. Izračunajte $D(𝑎,𝑏)$, $v(𝑎,𝑏)$ in $𝑎𝑏$ za pare celih števil $(2,5)$, $(4,10)$ in $(18,51)$. Kakšen je odnos med $D(a,b)$ in $v(a,b)$ v $ab$?
 
:::

::::{solution} ex-celaStevilaSageMath
:class: tip dropdown




In [87]:
divmod(956, 98)

(9, 74)

Količnik je 9 in ostanek je 74, ker $956 = 9 \cdot 98 + 74$.

--- 

In [88]:
234878 % 3 == 0

False

Rezultat je `False`, kar pomeni, da $3$ ne deli $234878$.

---


In [89]:
[divisors(n) for n in [134, 491, 422, 1002]]

[[1, 2, 67, 134],
 [1, 491],
 [1, 2, 211, 422],
 [1, 2, 3, 6, 167, 334, 501, 1002]]

Seznami deliteljev so:	
- Delitelji števila $134$ so $[1, 2, 67, 134]$.
- Delitelji števila $491$ so $[1, 491]$.
- Delitelji števila $422$ so $[1, 2, 211, 422]$.
- Delitelji števila $1002$ so $[1, 2, 3, 6, 167, 334, 501, 1002]$.
---
Med zgornjimi celimi števili je praštevilo samo $491$. To lahko ugotovimo tudi z naslednjim ukazom:


In [90]:
[is_prime(n) for n in [134, 491, 422, 1002]]

[False, True, False, False]

## Matematične konstante in funkcije


Pogoste konstante in funkcije so določene


$$ \sin(\pi) = 0$$


In [91]:
sin(pi)

0

In [92]:
cos(pi/4)

1/2*sqrt(2)

Vsak objekt v SageMath-u ima predstavitev LaTeX, ki jo dobimo z ukazom `latex()`


In [93]:
latex(cos(pi/4))

\frac{1}{2} \, \sqrt{2}

Če želimo da izhod je predstavljen v LaTeX-u, uporabimo čarobni ukaz `%display latex`

In [94]:
%display latex # Ponovno zaženite zgornje primere, potem ko izvedete ta ukaz.

Če želimo vrniti standardni izhod, uporabimo ukaz:

In [95]:
%display plain

V SageMath lahko delamo tudi z kompleksnimi števili. Imaginarno število $i$ je označeno z `i` in z `I`. Ker pa se `i` pogosto uporablja pri programiranju, je dobra praksa, da se označi z `I`.

In [96]:
arg(1+I)

1/4*pi

In [97]:
latex(1+I)

i + 1

Nekateri viri namesto $e$ raje pišejo $\exp$ in SageMath meni, da je to v redu.


In [98]:
e^3 - exp(3)

0

Ukaza `max()` in `min()` vrneta največje in najmanjše število v nizu števil.


In [99]:
max(1,2,3)

3

In [100]:
min(1,2,3)

1

V SageMathu, podobno kot v Pythonu, so seznami zapisani v oklepajih `[]`


In [101]:
max([2,3,4,5])

5

Ukaz `floor()` zaokroži število navzdol na najbližje celo število, ukaz `ceil()` pa zaokroži navzgor.


In [102]:
floor(2.1)

2

In [103]:
ceil(2.1)

3

**Pazite!**


In [104]:
floor(1/(2.1-2))

9

Ampak
$$ \left\lfloor \frac{1}{2.1 - 2} \right\rfloor = \left\lfloor \frac{1}{.1} \right\rfloor = \left\lfloor 10 \right\rfloor = 10$$

Kaj se je zgodilo?


Računalniki shranjujejo realna števila v binarni obliki, mi pa smo navajeni uporabljati desetiško predstavitev. Število $2.1$ v desetiškem zapisu je precej preprosto in kratko, vendar je po pretvorbi v binarni zapis $10.0\overline{0011} = 10.0001100110011 \dots$


Ker računalniki ne morejo shraniti neskončnega števila številk, se to število nekje zaokroži, kar povzroči manjšo napako, ki smo jo videli.
V SageMathu pa so racionalna števila (ulomki) natančna, zato te napake pri zaokroževanju nikoli ne bomo videli.


In [105]:
floor(1/ ( (21/10) - 2) )

10

:::{exercise}
:label: ex-matemKonstSageMath

1. Izračunajte celi del (oz. najbližje celo število) za število $e^{\pi}$.
2. Izračunajte logaritma z osnovo $e$ števila 1/1000000, izračunajte logaritma z osnovo $10$ števila 1/1000000, nato izračunajte razmerje. Kakšen naj bo odgovor?
3. Izračunajte logaritma z osnovo 2 števila 64.
4. Primerjaj $e^{i \pi}$ z njegovim numeričnim približkom z uporabo `pi.n()`.
5. Izračunajte $\sin(\pi/2)$, $\cot(0)$ in $\csc(\pi/16)$.

Rešitve napišite v ločeno celico besedila z uporabo LaTeXa.
:::


::::{solution} ex-matemKonstSageMath
:class: tip dropdown


In [106]:
floor(e**pi)

23

$$\left\lfloor e^{\pi} \right\rfloor = 23$$

log(1/1000000, e)

In [107]:
log(1/1000000, 10)	

-6

In [108]:
log(1/1000000, e) / log(1/1000000, 10)	

-1/6*log(1/1000000)


Razmerje je $ \log_{10}(e) $, kar je približno $0.434294481903252$.

---


log(64, 2)	


$\log_{2}(64) = 6$, ker je $2^6 = 64$.

---

In [109]:
e^(I*pi)

-1

In [110]:
e^(I*pi.n())

-1.00000000000000 + 1.22464679914735e-16*I

$$e^{i \pi} = -1$$ velja in je znana kot Eulerjeva identiteta. Numerični približek je zelo blizu $-1$, vendar ni natančno enak zaradi omejene natančnosti numeričnih izračunov.



::::

# Pripisovanje, enakost, enačbe in neenačbe.


SageMath uporablja `=` za pripisovanje. Za primerjavo uporablja `==`, `<=`, `>=`, `<` in `>`:


In [111]:
a=5
a

5

In [112]:
2 == 2

True

In [113]:
2<=3

True

In [114]:
2<2

False

In [115]:
a < 3

False

In [116]:
a == 5

True

Če želimo shraniti rezultat izračuna, ga dodelimo spremenljivki:


In [117]:
y=1+2

**Pazite**: Pogosta napaka je uporaba izraza `a=b` za primerjavo `a` in `b`, vendar na koncu vrednost `a` zamenjate z vrednostjo `b`.


In [118]:
y*(y+2)

15

Rezultat izračuna se ne natisne samodejno, ko je pripisan spremenljivki.
Zato bomo naredili naslednje, da ga tudi natisnemo,


In [119]:
b=3*4 ; b

12

Znak `;`, ki ločuje več navodil v isti vrstici.
Če ni določeno drugače, se v beležnici Jupyter natisne samo izpis, ki ga ustvari zadnja vrstica. Zato je uporaba znaka `;` enakovredna uporabi nove vrstice v isti celici.


In [120]:
y = 3 + 2; y
y+2

7

Če želimo natisniti vsak korak, moramo uporabiti ukaz `print()`


In [121]:
y = 1 + 2; print(y)
y = 3 * y + 1; print(y)
y = 3 * y + 1; print(y)
y = 3 * y + 1; print(y)

3
10
31
94


- Če znate programirati v Pythonu, boste morda ugotovili, da je to standardno obnašanje spremenljivke.


- Ni priporočljivo na novo definirati vnaprej določenih konstant in funkcij iz programa Sage.
- Čeprav to ne vpliva na notranje obnašanje programa Sage, lahko privede do presenetljivih rezultatov:


In [122]:
pi = -I/2

In [123]:
exp(2*I*pi)

e

Ukaz `restore()` povrne privzeto vrednost vsem vnaprej določenim spremenljivkam in funkcijam.


In [124]:
restore()

In [125]:
y+pi

pi + 94

Ukaz `reset()` opravi še popolnejšo ponastavitev, zlasti izbriše vse uporabniško definirane spremenljivke.


In [126]:
reset()

In [127]:
y

NameError: name 'y' is not defined

Opazite tudi, da SageMath včasih razlikuje med simbolnimi izračuni in števili.


In [ ]:
e^(2*pi*I)

In [ ]:
e^(2*pi*I) == 1

SageMath zgornjega ne razume kot primerjavo, temveč kot enačbo.


In [ ]:
3*x - 10 == 5

Enačbo ali neenakost rešimo z ukazom `solve()`.


In [ ]:
solve(3*x - 2 == 5,x)

Privzeto SageMath interpretira `x` kot simbolno spremenljivko. Za zdaj bomo reševali samo `x`, pozneje pa bomo videli, kako določiti lastne (simbolne) spremenljivke


In [ ]:
solve( 2*x - 5 >= 17,x)

In [ ]:
solve( x^2 + x == 6, x)

In [ ]:
solve(2*x^2 - x + 1 == 0, x)

Množica rešitev nekaterih neenačb je sestavljena iz unije in preseka odprtih intervalov.


In [ ]:
solve( x^2 - 6 >= 3, x )

In [ ]:
solve( x^2 - 6 <= 3, x )

- Ukaz `solve()` poskuša izraziti rešitev enačbe brez uporabe števil s plavajočo vejico.
- Če tega ni mogoče storiti, bo vrnil rešitev v simbolični obliki.


In [ ]:
solve( sin(x) == x, x)

- Za iskanje numeričnega približka rešitve lahko uporabimo ukaz `find_root()`.
- Ta ukaz zahteva izraz in zaprt interval, na katerem se išče rešitev


In [ ]:
find_root(sin(x) == x, -pi/2 , pi/2)

- Ob začetku seje SageMath ustvari eno simbolno spremenljivko, `x`, ki se lahko uporablja za reševanje enačb.
- Če želite uporabiti dodatno simbolno spremenljivko, jo morate deklarirati z ukazom `var()`.


In [ ]:
y,z,t = var("y z t")

In [ ]:
phi, theta, rho = var("phi theta rho")

In [ ]:
x1, x2 = var("x1 x2")

- Imena spremenljivk ne smejo vsebovati presledkov
- Na primer `square root` ni veljavno ime spremenljivke, `square_root` pa je.


Poskus uporabe simbolne spremenljivke, preden je bila deklarirana, bo povzročil napako `NameError`.


In [ ]:
u

In [ ]:
solve (u^2-1,u)

Simbolno spremenljivko, kot je zgoraj opredeljena spremenljivka `phi`, lahko razglašamo z ukazom `restore()`.


In [ ]:
restore('phi')

In [ ]:
phi

Majhne sisteme linearnih enačb lahko rešite tudi s ukazom `solve()`, če imajo vse simbolne spremenljivke so bile deklarirane.


In [ ]:
x,y,z = var('x','y','z')

In [ ]:
solve( [3*x - y == 2,-2*x -y == 1 ], x,y)

[[x == (1/5), y == (-7/5)]]

In [ ]:
solve( [ 2*x - y == -1 , 2*x - y == 2],x,y)

In [ ]:
solve( [ 2*x + y == -1 ,-4*x - 2*y == 2],x,y)

- `r1` označuje, da obstaja prosta spremenljivka, ki določa množico rešitev.


- Če je več prostih spremenljivk, jih SageMath našteje `r1,r2,..., rk`.


In [ ]:
solve([ 2*x + 3*y + 5*z == 1, 4*x + 6*y + 10*z == 2, 6*x + 9*y + 15*z == 3], x,y,z)

[[x == -5/2*r3 - 3/2*r4 + 1/2, y == r4, z == r3]]

- Ukaz `solve()` je lahko pri velikih sistemih enačb zelo počasen.
- Za te sisteme je najbolje uporabiti funkcije linearne algebre, saj so precej učinkovite. Kmalu se bomo naučili, kako.


Reševanje neenačb v več spremenljivkah lahko privede do zapletenih izrazov, saj so območja, ki jih določajo, zapletena.


In [ ]:
solve([ x-y >=2, x+y <=3], x,y)

### Vaje

1. Poiščite vse rešitve enačbe $x^{3} - x = 7x^{2}-7$
2. Poišči celotno množico rešitev za neenačbo $|t-7| \geq 3$
3. Poiščite vsa $x$ in $y$, ki izpolnjujejo tako $2𝑥+ 𝑦= 17$ kot $𝑥-3𝑦=-16$.
4. S funkcijo `find_root()` poiščite rešitev enačbe $e^{x} = \cos(𝑥)$ na intervalu $[-\pi/2,0]$.
5. Zgornji ukaz spremenite tako, da find_root()najde drugo rešitev na istem intervalu.


# Lepe praznike!


In [ ]:
var('t k')
from sage.plot.plot3d.shapes import *
num_wraps=4
num_lights=30
c=rainbow(num_lights)
f(t) = ( (1-t)*cos(2*pi*t*num_wraps), (1-t)*sin(2*pi*t*num_wraps),t)
lights=add([point(f(1/(num_lights)*(k+1)),color=c[k],size=50) for k in range(0,num_lights)])
S = Cone(1, 1, color='green')
S += parametric_plot3d(f(t),(t,0,1),thickness=3  )
S += lights
S.show(frame=false)

Graphics3d Object